CodeModel/CodeParsers/CSharp/ClassCohesionAnalyzer/CohesionAnalyzer.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using CodeModel.CaDETModel.CodeItems;
+
+namespace CodeModel.CodeParsers.CSharp.ClassCohesionAnalyzer
+{
+    public class CohesionAnalyzer
+    {
+        public List<CohesiveParts> IdentifyCohesiveParts(CaDETClass parsedClass)
+        {
+            ClassInteractions classInteractions = new ClassInteractions(parsedClass);
+
+            IEnumerable<CohesiveParts> possibleParts = GetAllPossibleParts(classInteractions);
+
+            return possibleParts.Where(cohesiveParts =>
+                    cohesiveParts.Parts.TrueForAll(set => CalculateLCOM(set) < 0.4))
+                .ToList();
+        }
+
+        private IEnumerable<CohesiveParts> GetAllPossibleParts(ClassInteractions classInteractions)
+        {
+            var accessesThatCannotBeCut = classInteractions.GetAccessesThatCannotBeCut();
+            var accessesToBeCut = GetAccessesThatCanBeCut(classInteractions.Accesses, accessesThatCannotBeCut);
+            var possibleParts = new List<CohesiveParts>();
+            foreach (var accesses in accessesToBeCut)
+            {
+                if (possibleParts.Any(s => accesses.IsProperSupersetOf(s.AccessesToCut))) continue;
+                var parts = FindValidCohesiveParts(accesses, new HashSet<Access>(classInteractions.Accesses));
+                if (parts is not { Count: 2 }) continue;
+                var cohesiveParts = new CohesiveParts(accesses, parts);
+                possibleParts.RemoveAll(s => s.AccessesToCut.IsProperSupersetOf(cohesiveParts.AccessesToCut));
+                possibleParts.Add(cohesiveParts);
+            }
+
+            return possibleParts;
+        }
+
+        private IEnumerable<HashSet<Access>> GetAccessesThatCanBeCut(HashSet<Access> allAccesses,
```

Comment:
I find it a bit odd that the methods GetAccessesThatCanBeCut and GetAccessesThatCannotBeCut are in separate classes.